### PR TITLE
fix(binder): castability check between array and string

### PR DESCRIPTION
**Pull request by xiangjinwu**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

> The automatic casts to string types are treated as assignment casts, while the automatic casts from string types are explicit-only.
 
https://www.postgresql.org/docs/14/sql-createcast.html#id-1.9.3.58.7.4

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)

#4831

### CODE REVIEW

**Comment by xiangjinwu** [comment]
@nanderstabel Could you please explain the original logic with some examples?

**Comment by nanderstabel** [comment]
> @nanderstabel Could you please explain the original logic with some examples?

Originally I only added this arm to check for casting compatibility:
```
        (
            DataType::Varchar,
            DataType::List {
                datatype: target_elem,
            },
        ) => cast_ok(&DataType::Varchar, target_elem, allows),
```
In hindsight, I see now that it is unnecessary to continue to checking for casting compatibility here to the `target_elem` type.

 The reason I added:
 ```
        (
            DataType::Varchar,
            DataType::List {
                datatype: target_elem,
            },
        ) if target_elem == &Box::new(DataType::Varchar) => true,
```

was because of this:
```
dev=> select arr::varchar[] from (values ('{1, 2, 3}')) as t(arr);
ERROR:  QueryError: Bind error: cannot cast type "varchar" to "varchar[]" in Explicit context
```

Not realizing that simply 
```(DataType::Varchar, DataType::List { datatype: _ }) => CastContext::Explicit <= allows,```
would suffice for both match arms.

**Comment by xiangjinwu** [comment]
Thanks for the example. Shall we proceed with the new implement then? If yes, I will polish this PR with some tests.

**Comment by nanderstabel** [comment]
Yes please proceed with your implementation, it's much cleaner!

**Comment by nanderstabel** [comment]
btw this PR is related to this issue: #5063

It will fix casting Varchar to multidimensional List. You may want to wait for it to be merged (in a bit) before writing tests.

File: src/expr/src/vector_op/cast.rs [hunk 1]
```diff
@@ -268,6 +269,7 @@ macro_rules! for_each_cast {
             { date, varchar, general_to_string },
             { timestamp, varchar, general_to_string },
             { timestampz, varchar, timestampz_to_utc_string },
+            { list, varchar, |x| general_to_string(x) },
```
Comment (richardchien) [inline src/expr/src/vector_op/cast.rs:272]: What's the difference between `general_to_string` and `|x| general_to_string(x)`?

**Review by richardchien — APPROVED** [review]
Rest LGTM!

File: src/expr/src/vector_op/cast.rs (hunk 1 above)
Comment (xiangjinwu) [reply src/expr/src/vector_op/cast.rs:272]: It is like the relationship between `min` and `min_str` in batch agg:
https://github.com/risingwavelabs/risingwave/blob/c343684d35347f5cdeea1a3e9f2c4acaecfa6561/src/expr/src/vector_op/agg/functions.rs#L85-L87

When the compiler fills the generic function with a concrete type, a fixed lifetime is chosen at the same time. However, we need a function that works for any lifetime (`for<'a>`).

**Pull request merged**